### PR TITLE
QVM: Fix bitrot and add CI.

### PR DESCRIPTION
**Pull request by dsvensson**

### CODE REVIEW

**Comment by dsvensson** [comment]
Not sure how you want to do the CI. Would be nice if ktx also produced GH downloads for all builds, also in fork repos. It's free and convenient.

**Comment by ciscon** [comment]
if you're logged in you can download the artifacts from the action.

**Comment by dsvensson** [comment]
yeah, but I'm thinking of this `if: github.repository == 'QW-Group/ktx'` in `build-targets.yml` and `build-targets-snapshot.yml`.

**Comment by ciscon** [comment]
the snapshots get pushed, but yeah definitely the targets one should be everything, on ezquake it's: on: [push, pull_request, workflow_dispatch]

**Comment by dsvensson** [comment]
Skärmavbild 2024-06-10 kl  20 12 02

**Comment by ciscon** [comment]
the other problem is just that we don't have an upload artifact in the target one (or the others, though they have a different purpose anyway).  i was going to combine all of them into one and just have different logic paths as right now we're doing the same work multiple times but have yet to get to it.

File: src/g_syscalls_extra.c [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+	return ival;
+}
+
+float trap_GetExtField_f(gedict_t *ed, const char *fieldname)
+{
+	fi_t tmp;
+	tmp._float = -1.0f;
+	if (HAVEEXTENSION(G_GETEXTFIELD))
+	{
+		tmp._int = trap_GetExtField_f(ed, fieldname);
```
Comment (dsvensson) [inline src/g_syscalls_extra.c:52]: infinite recursion copypasta 🙄

**Comment by dsvensson** [comment]
If you don't have time to integrate the CI changes properly in the other builds feel free to suggest how you want it done and I'll fix it. Would be great if the bitrot could be stopped once and for all before more comes creeping in.

**Comment by ciscon** [comment]
the goal would be to have a single workflow that always runs the logic in the current build-target.yml (but then uploads those as artifacts) and then just like the other separate steps currently have push those as snapshots if it's a push on qw-group/ktx and do a release with them if it's a release.  i was going to reorganize it so there was one main file with the logic and three separate ones to be included with the other logic, but all in one file would also be just fine as once it's deduplicated it really won't be much longer than any one of those files currently (and we'll only be building the once).

**Comment by dsvensson** [comment]
https://builds.quakeworld.nu/ktx is not even up, why not just go with github artifacts?

**Comment by ciscon** [comment]
that's exactly what i'm saying, the artifacts are only created for one of the pipelines right now and not reused- they should be created regardless and then reused depending on X for either pushing snapshots and/or release instead of rebuilding things.  we do want a place for the snapshot stuff though and not just do drafts/releases for every single build, that doesn't mean there shouldn't be artifacts attached to the build though.

**Comment by dsvensson** [comment]
I guess the scp stuff could remain. I'll hack up something that works with the github release flow, and that could also copy the release artifacts elsewhere.

**Comment by ciscon** [comment]
it's actually pretty simple, we've just got it all divided up because it was done multiple times- always do the build creating artifacts, then only do the snapshot push of those artifacts (fiddled with for naming etc) if it's the right repo/branch, and only do the push of the artifacts to release on release/published.

**Comment by dsvensson** [comment]
Almost done now, releases end up at your equivalence to:

https://github.com/qw-ctf/ktx/releases

...like other GitHub projects. On that page you see a _'Draft a new release'_ that gives you this:
image

After entering a new release you generate release notes that will auto-populate based on pull requests merged since last release.

This will trigger the build including the upload job that will attach release artifacts to the release, as well as upload to builds.quakeworld.nu.

Here's the current incarnation:
https://github.com/qw-ctf/ktx/blob/arriba/.github/workflows/main.yml

Think I just need to place the final `if:` guards, and perhaps trim some noisy output.

**Comment by ciscon** [comment]
i haven't looked at it yet but does it push a snapshot when it's merged into master on qw-group or after a release is tagged?

**Comment by dsvensson** [comment]
When the `if:` guards are finalized it will push snapshots on master pushes, and releases on release events.

**Comment by dsvensson** [comment]
@qqshka @tcsabina Added build timeout per feedback from @vikpe, and @ciscon was the one suggesting the CI changes, so do you have any further comments?

**Comment by tcsabina** [comment]
> @qqshka @tcsabina Added build timeout per feedback from @vikpe, and @ciscon was the one suggesting the CI changes, so do you have any further comments?

Hi Nano,

let me get back to you within a few days about this CI/CD change...

**Comment by dsvensson** [comment]
Verified that https://builds.quakeworld.nu/ktx/snapshots/ is populated correctly.

Don't think there's any need to test release path specifically as it uses the same code.

Here's what it looks like when building in my branch where it just echoes the result:
https://github.com/qw-ctf/ktx/actions/runs/9820556161/job/27115280956

**Comment by ciscon** [comment]
looks good to me

**Pull request merged**